hasher.go

```diff
@@ -0,0 +1,130 @@
+package smt
+
+import (
+	"bytes"
+	"hash"
+)
+
+var (
+	leafPrefix  = []byte{0}
+	innerPrefix = []byte{1}
+	extPrefix   = []byte{2}
+)
+
+var _ PathHasher = (*pathHasher)(nil)
+var _ ValueHasher = (*valueHasher)(nil)
+
+// PathHasher defines how key inputs are hashed to produce tree paths.
+type PathHasher interface {
+	// Path hashes a key (preimage) and returns a tree path (digest).
+	Path([]byte) []byte
+	// PathSize returns the length (in bytes) of digests produced by this hasher.
+	PathSize() int
+}
+
+// ValueHasher defines how value data is hashed to produce leaf data.
+type ValueHasher interface {
+	// hashValue hashes value data to produce the digest stored in leaf node.
+	hashValue([]byte) []byte
+}
+
+type treeHasher struct {
+	hasher    hash.Hash
+	zeroValue []byte
+}
+type pathHasher struct {
+	treeHasher
+}
+type valueHasher struct {
+	treeHasher
+}
+
+func newTreeHasher(hasher hash.Hash) *treeHasher {
```

Comment:
Publically exposing `NewTreeHasher` and also adding a `NewPathHash` would be helpful for upstream libraries using this.

Reply:
the `treeHasher` interface is internal-only, and only the path/value hashers are exposed to keep the concerns separate. So I could expose constructors for those, but the interfaces are trivial anyway.